### ISSUE TITLE
MSVC 2017 and define PYBIND11_CPP17

### ISSUE THREAD

**Issue by RedSkotina** [pybind/pybind11#833]
PYBIND11_CPP17 not defined for msvc 2017 but msvc 2017 has support for extended constexpr and std::variant with compiler option `/std:c++latest`
I try explicit define PYBIND11_CPP17 on msvc and run test succesufull. (as far as I understand)

[https://github.com/pybind/pybind11/blob/master/include/pybind11/common.h#L19](url)
```
// Neither MSVC nor Intel support enough of C++14 yet (in particular, as of MSVC 2015 and ICC 17
// beta, neither support extended constexpr, which we rely on in descr.h), so don't enable pybind
// CPP14 features for them.
#if !defined(_MSC_VER) && !defined(__INTEL_COMPILER)
#  if __cplusplus >= 201402L
#    define PYBIND11_CPP14
#    if __cplusplus > 201402L /* Temporary: should be updated to >= the final C++17 value once known */
#      define PYBIND11_CPP17
#    endif
#  endif
#endif
```
Can you add PYBIND11_CPP17  support for msvc 2017  (with and without  `/std:c++latest` )?

**Comment by jagerman** [comment]
If what you want is support for `stl.h` enabling `<optional>` and `<variant>` under MSVC 2017, a patch for that sounds pretty reasonable.  Are these always available, or does availability depend on a particular `/std:` flag?

As an aside, we can't reasonably define `PYBIND11_CPP17`, until MSVC supports considerably more of C++17 (for the most part, MSVC 2017's C++17 support is limited to stl features and almost devoid of core language features).  We can't even set `PYBIND11_CPP14`: MSVC still chokes on the C++14 codepath in `descr.h`.  While it claims to support relaxed constexpr, it doesn't seem to support it robustly: the combination of relaxed constexpr and `constexpr auto` return types in `descr.h` doesn't work.  I did try turning on it when MSVC 2017 came out; it failed, and so the define is still off.

**Comment by RedSkotina** [comment]
some info:
1.  

> /std:c++14
> The /std:c++14 switch enables the set of C++14 features implemented by the Visual C++ compiler, modulo conformance bug fixes and ongoing feature work (such as for constexpr, expression SFINAE, 2-phase name lookup, etc.) Using the /std:c++14 switch will always instruct Visual C++ to compile your code according to the C++14 standards and associated Defect Reports.  This is the default language version in this update.

> /std:c++latest
> The /std:c++latest switch will move forward the set of features implemented by Visual C++ with each new version of the C++ standards Working Draft. Currently it will track features that we add for the proposed post-C++14 draft standard. When the next standard is approved, the semantics opted-in with the /std:c++latest will be covered by /std:c++17 (assuming that the next standard is C++17, but the standards committee working group never ceases to surprise) and /std:c++latest will move forward to track the next draft standard’s features. In another words, with /std:c++latest, you will always get the set of all new features currently implemented and on by default.

Now `std::variant`, `std::optional`, `std::any` available as part of c++17 ONLY with /std:c++latest. Later it must be available with `/std:c++17` (we dont have such parameter now)
How we can detect this parameters ?

> _MSVC_LANG Defined as an integer literal that specifies the C++ language standard targeted by the compiler. When compiled as C++, the macro is the integer literal value 201402 if the /std:c++14 compiler option is set, or by default, and it is set to a higher, unspecified value when the /std:c++latest compiler option is set. Otherwise, the macro is undefined. The _MSVC_LANG macro and /std (Specify Language Standard Version) compiler options are available beginning in Visual Studio 2015 Update 3.

2. I try explicit define PYBIND11_CPP14 with /std:c++latest . It is fail, as you say with partitial constexpr support. 
 But with explicit define PYBIND11_CPP17 with /std:c++latest  and PYBIND11_HAS_VARIANT i can compile simple example with variant. 
to be continue...

**Comment by jagerman** [comment]
#841 should address this, by setting `CPP14` for MSVC and, if you use /std:c++latest under MSVC 2017, `CPP17`.  It appears to work (and passes the optional/variant tests), but give it a try and confirm that it works for you.

**Comment by RedSkotina** [comment]
ok. must i test it before merge pullrequest in master or after ?

**Comment by jagerman** [comment]
Fixed via #841.
